cpp/tests/experimental/spatial/pairwise_point_in_polygon_test.cu

```diff
@@ -0,0 +1,382 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/iterator_factory.cuh>
+#include <cuspatial/experimental/pairwise_point_in_polygon.cuh>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/device_vector.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <gtest/gtest.h>
+
+using namespace cuspatial;
+
+template <typename T>
+struct PairwisePointInPolygonTest : public ::testing::Test {
+ public:
+  rmm::device_vector<vec_2d<T>> make_device_points(std::initializer_list<vec_2d<T>> pts)
+  {
+    return rmm::device_vector<vec_2d<T>>(pts.begin(), pts.end());
+  }
+
+  rmm::device_vector<std::size_t> make_device_offsets(std::initializer_list<std::size_t> pts)
+  {
+    return rmm::device_vector<std::size_t>(pts.begin(), pts.end());
+  }
+};
+
+// float and double are logically the same but would require separate tests due to precision.
+using TestTypes = ::testing::Types<float, double>;
+TYPED_TEST_CASE(PairwisePointInPolygonTest, TestTypes);
+
+TYPED_TEST(PairwisePointInPolygonTest, OnePolygonOneRing)
+{
+  auto point_list        = std::vector<std::vector<float>>{{-2.0, 0.0},
+                                                    {2.0, 0.0},
+                                                    {0.0, -2.0},
+                                                    {0.0, 2.0},
+                                                    {-0.5, 0.0},
+                                                    {0.5, 0.0},
+                                                    {0.0, -0.5},
+                                                    {0.0, 0.5}};
+  auto poly_offsets      = this->make_device_offsets({0});
+  auto poly_ring_offsets = this->make_device_offsets({0});
+  auto poly_point =
+    this->make_device_points({{-1.0, -1.0}, {1.0, -1.0}, {1.0, 1.0}, {-1.0, 1.0}, {-1.0, -1.0}});
+
+  auto got      = rmm::device_vector<int32_t>(1);
+  auto expected = std::vector<int32_t>{false, false, false, false, true, true, true, true};
+
+  for (size_t i = 0; i < point_list.size(); ++i) {
+    auto point = this->make_device_points({{point_list[i][0], point_list[i][1]}});
+    auto ret   = pairwise_point_in_polygon(point.begin(),
+                                         point.end(),
+                                         poly_offsets.begin(),
+                                         poly_offsets.end(),
+                                         poly_ring_offsets.begin(),
+                                         poly_ring_offsets.end(),
+                                         poly_point.begin(),
+                                         poly_point.end(),
+                                         got.begin());
+    EXPECT_EQ(got, std::vector<int32_t>({expected[i]}));
+    EXPECT_EQ(ret, got.end());
+  }
```

Comment:
Ah ok. I see how you are computing these results 1 by 1. I think you can wrap the iterators of input point arrays with [thrust::permulation_iterator](https://nvidia.github.io/thrust/api/classes/classthrust_1_1permutation__iterator.html#class-thrustpermutation_iterator) and set the offsets accordingly to make 8 polygons without duplicate code.